neuland/simulation/generator/CosmicMuon.h

```diff
@@ -0,0 +1,137 @@
+    {
+      public:
+        TrackGenerator(AngleDist angle_dist, EnergyDist energy_dist, PointDist point_dist)
+            : angle_dist_{ angle_dist }
+            , energy_dist_{ energy_dist }
+            , point_dist_{ point_dist }
+        {
+        }
+
+        void set_detector_size(double detector_size) { detector_size_ = detector_size; }
+        void set_rd_engine(TRandom* user_rd_engine) { rd_engine_ = user_rd_engine; }
+
+      private:
+        static constexpr auto CLight = ::Neuland::CLight;
+        double detector_size_{ 50.0 };
+
+        AngleDist angle_dist_{};
+        EnergyDist energy_dist_{};
+        PointDist point_dist_{};
+        TRandom* rd_engine_{ gRandom };
+
+        auto rd_num_gen_point() -> ROOT::Math::Cartesian3D<double>;
+        auto rd_num_gen_angles(AngleDist angle_dist_) -> ROOT::Math::Polar3D<double>;
+        auto rd_num_gen_energy() -> double;
+        auto calculate_abs_momentum(double kinetic_energy) -> double { return kinetic_energy / CLight; };
+        auto calculate_momentum_energy(double kinetic_energy, AngleInfo& angle_info) -> ROOT::Math::PxPyPzE4D<double>;
+
+        auto calculate_external_position_momentum(PointDist point_dist_, AngleDist angle_dist_, EnergyDist energy_dist_)
+            -> std::pair<ROOT::Math::PxPyPzE4D<double>, ROOT::Math::Cartesian3D<double>>;
+
+        auto ReadEvent(FairPrimaryGenerator* prim_gen) -> Bool_t override
+        {
+
+            auto position_momentum = std::pair<ROOT::Math::PxPyPzE4D<double>, ROOT::Math::Cartesian3D<double>>{
+                calculate_external_position_momentum(point_dist_, angle_dist_, energy_dist_)
+            };
+            prim_gen->AddTrack(13,
+                               position_momentum.first.Px(),
+                               position_momentum.first.Py(),
+                               position_momentum.first.Pz(),
+                               position_momentum.second.X(),
+                               position_momentum.second.Y(),
+                               position_momentum.second.Z());
+
+            return true;
+        };
+    };
+
+    template <typename AngleDist, typename EnergyDist, typename PointDist>
+    auto TrackGenerator<AngleDist, EnergyDist, PointDist>::rd_num_gen_angles(AngleDist angle_dist)
+        -> ROOT::Math::Polar3D<double>
+    {
+        auto angles = ROOT::Math::Polar3D<double>{};
+        angles.SetPhi(rd_engine_->Uniform(0., M_PI));
+        angles.SetTheta(angle_dist(rd_engine_));
+        return angles;
+    }
+
+    template <typename AngleDist, typename EnergyDist, typename PointDist>
+    auto TrackGenerator<AngleDist, EnergyDist, PointDist>::calculate_momentum_energy(double kinetic_energy,
+                                                                                     AngleInfo& angle_info)
+        -> ROOT::Math::PxPyPzE4D<double>
+    {
+        auto momentum_energy = ROOT::Math::PxPyPzE4D<double>{ 0, 0, 0, kinetic_energy };
```

Comment:
Also a type alias in the PRIVATE section:
```c++
using Momentum =  ROOT::Math::PxPyPzE4D<double>;
```
```suggestion
        auto momentum_energy = Momentum{ 0, 0, 0, kinetic_energy };
```